File: Swift 풀이/Graph/[백준 11403] 경로 찾기.swift
//
//  [백준 11403] 경로 찾기.swift
//  Algorithm
//
//  Created by 김민 on 5/1/24.
//
// 백준 11403 경로 찾기 https://www.acmicpc.net/problem/11403

/*
[📌 전략]
 그래프 문제

[✅ 풀이]
 행으로 나타내진 각 정점을 모두 방문하여 정답을 구한다.
 한 정점 기준으로 인접 노드가 1이면 방문 처리를 하고 dfs(인접 노드)를 수행한다.
 이때 visited 배열을 전체 행렬에 대해서가 아닌 행 하나를 기준으로 만들어 시작 노드가 해당 노드로 갈 수 있는 길이 있는지를 구한다.

[📝 기록]
 어려운 문제가 아닌 게 분명한데 혼자 풀이하지 못했다. dfs 개념을 더 자세하게 봐야겠음.
*/


let n = Int(readLine()!)!
var matrix = Array(repeating: Array(repeating: 0, count: n), count: n)
var visited = Array(repeating: false, count: 102)

for i in 0..<n {
    matrix[i] = readLine()!.split(separator: " ").map { Int($0)! }
}

func dfs(_ v: Int) {
    for i in 0..<n {
        if !visited[i] && matrix[v][i] == 1 {
            visited[i] = true
            dfs(i)
        }
    }
}

for i in 0..<n {
    visited =  Array(repeating: false, count: 102)
    dfs(i)
    for j in 0..<n {
        if visited[j] { matrix[i][j] = 1 }
    }
}

matrix.forEach {
    print($0.map { String($0) }.joined(separator: " "))
}

File: Python/그래프/[백준 11403] 경로 찾기.py
'''
[백준 11403] 경로 찾기
https://www.acmicpc.net/problem/11403

[📌 전략]
인접 행렬 형식으로 주어진 그래프를 탐색하는 문제

[✅ 풀이]
1. 행렬을 사용해서 단방향 그래프 입력 처리
2. dfs 함수
    - 파라미터: cur(현재 노드), st(시작 노드)
    - cur의 인접 노드를 순회하며 board[st][nxt]가 방문 전이면 1로 갱신 후 해당 노드에서 dfs 탐색
    (board 배열을 visited 배열과 겸하여 사용
3. 0~n 순회하며 dfs 함수 호출

[📝 기록]
LV: 실버1
풀이 시간: 39분
'''

import sys

input = sys.stdin.readline

n = int(input())
g = [[] for _ in range(n)]
board = [[0 for _ in range(n)] for _ in range(n)]

# 행렬 입력 처리
for i in range(n):
    line = list(map(int, input().split()))

    for j in range(n):
        if line[j] == 1: g[i].append(j)

# dfs 함수
def dfs(cur, st):
    for nxt in g[cur]:
        if board[st][nxt] == 1: continue

        board[st][nxt] = 1
        dfs(nxt, st)

# dfs 탐색 수행
for i in range(n): dfs(i, i)

# 답안 출력
for i in range(n): print(*board[i])